src/metabase/task/sync_databases.clj

```diff
@@ -254,37 +256,30 @@
-  The defaults used to be all at the same time leading to poor resource management. If the user has not indicated they
-  want to control the scheduling, returns appropriate randomized schedules for the sync tasks."
-  [{:keys [details cache_field_values_schedule metadata_sync_schedule] :as _database}]
-  (when-not (:let-user-control-scheduling details)
-    (let [random-defaults (sync.schedules/schedule-map->cron-strings (sync.schedules/default-schedule))
-          old-default-keys (cond-> []
-                             (contains? sync.schedules/default-cache-field-values-schedule-cron-strings cache_field_values_schedule)
-                             (conj :cache_field_values_schedule)
-
-                             (contains? sync.schedules/default-metadata-sync-schedule-cron-strings metadata_sync_schedule)
-                             (conj :metadata_sync_schedule))]
-      (not-empty (select-keys random-defaults old-default-keys)))))
-
-(defn maybe-update-db-schedules
-  "Update schedules when managed by metabase to spread the syncs out in time.
-  When not let-user-control-scheduling metabase is in control of the sync times. These used to all be scheduled at
-  every hour at 50 minutes or on the hour. This can cause a large load on the databaes or CPU and this attempts to
-  spread them out in time."
+(defn- metabase-controls-schedule?
+  "Predicate returning if the user does not manually set sync schedules and leaves it to metabase."
   [database]
-  (if-let [randomized-schedules (randomized-schedules database)]
-    (u/prog1 (merge database randomized-schedules)
-      (db/update! Database (u/the-id database) randomized-schedules))
-    database))
+  (not (-> database :details :let-user-control-scheduling)))
+
+(defn- old-default-schedule?
+  "Database uses the old default scheduled time for syncs. All databases used to sync schema on the hour and sync field
+  values at midnight. This led to stampedes on db resources so we now randomize the schedules by default."
+  [database]
+  (or (contains? sync.schedules/default-cache-field-values-schedule-cron-strings
+                 (:cache_field_values_schedule database))
+      (contains? sync.schedules/default-metadata-sync-schedule-cron-strings
+                 (:metadata_sync_schedule database))))
 
 (defmethod task/init! ::SyncDatabases
   [_]
   (job-init)
-  (doseq [database (db/select Database)]
-    (try
-      (check-and-schedule-tasks-for-db! (maybe-update-db-schedules database))
-      (catch Throwable e
-        (log/error e (trs "Failed to schedule tasks for Database {0}" (:id database)))))))
+  ;; prevent "stampedes" on databases which all used the same sync schedules. Randomize them
+  (let [dbs-with-default-schedules (->> (db/select Database)
+                                        (filter
+                                         (every-pred
+                                          metabase-controls-schedule? old-default-schedule?)))]
```

Comment:
It seems super inefficient to fetch the entire row for _every_ Database on every launch like this. You could extremely easily only select the fields you need, e.g.

```clj
(db/select [Database :id :details ...])
```

You can also extremely easily filter do `old-default-schedule?` directly in SQL with something like

```clj
(db/select [Database :id :details ...]
           {:where [:or
                    [:in :cache_field_values_schedule default-cache-field-values-schedule-cron-strings]
                    [:in :metadata_sync_schedule default-metadata-sync-schedule-cron-strings]]})
```

Unfortunately `old-default-schedule?` still has to be done in Clojure-land but once #22354 gets merged (and it seems like [@bshepherdson](https://github.com/bshepherdson) is really close on getting it ready) we can do the entire thing in Database-land. At which point in the normal happy-path `db/select` would return no rows.

The only reason I'm mentioning this is there can be lots of Databases in big instances and fetching 1000 entire rows over the wire and doing all the Toucan magic can be a little expensive. Doing things in the Database is way faster so if we can do it there we should

Reply:
I would also suggest using `db/select-reducible` and reducing/transducing the results so we're handling each row as it comes and then tossing it instead of potentially fetching 1000 rows at once and then iterating over all of them. That should make this faster and reduce the memory impact

Reply:
I'm 100% on board with this. I'm not following why `old-default-schedule?` cannot be done in database land given that you gave an example above it. I think you're talking json access for `:let-user-control-scheduling` check for `metabase-controls-schedule?`? If so I'm on board with these changes.

Reply:
Oh yeah sorry that's what I meant. `let-user-control-scheduling` can't be done in Clojure-land until we get the H2 upgrade in so we have native JSON functions. 

Well actually if you wanted to be hacky I suppose it could be done in Database-land for Postgres and MySQL and in Clojure-land for H2 until #22354 gets merged in. Maybe a bit too much work. It would be nice to add a comment tho or open a GH issue so we remember to come change it over later